File: 98-validate-binary-search-tree/98-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def isValid(root, ub, lb):
            # Base case. None is True --> every leaf node is a valid BST tree
            if root is None: return True
            
            if ub is not None and root.val >= ub.val or lb is not None and root.val <= lb.val:
                return False
        
            return isValid(root.left, root, lb) and isValid(root.right, ub, root)
        
        return isValid(root, None, None)

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
    function isValid(root, ub, lb) {   
        return root == null || (ub == null || root.val < ub.val) && (lb == null || root.val > lb.val) && isValid(root.left, root, lb) && isValid (root.right, ub, root)
    }
    
    return isValid(root, null, null)
};